### DISCUSSION TITLE
Tailoring /Reducing Flash consumption u8g2 lib

### DISCUSSION THREAD

**Discussion by RobLeop** [olikraus/u8g2#2070] [General]
Hello!
At first: Great Lib! It was really easy and well explained to create in a short time nice screens with the Lib functions.

Are there some tips to tailor the lib?
Are already not used functions optimized by the arduino compiler? Or can I safe some flash by commenting some functions /definitions?

The Lib and functions I used consumed more than 10kB of flash.
I already used the smallest font libs of the choosen style I need (e.g. sometimes just numbers).
Maybe you have some other tips for me below I printed a regular function I use to create different screens for my project.
Are some fucntions really need a lot of flash (e.g. get the character width / height...)

Thanks a lot in advance!

Usual functions I use:

oled.firstPage();
  do
  { 
    // Draw Caption with Line and little Logo   
    oled.setFont(u8g2_font_helvB10_tr);
    oled.setFontDirection(0); // text is horizontal
    oled.setFontPosTop(); // text is build up from top of the font (top line)
    oled.setCursor(1,1);
    oled.print(caption);
    oled.drawLine(0, oled.getMaxCharHeight()+2,ScreenWidth_lastpixel, oled.getMaxCharHeight()+2); // draw a line below the "caption" of the screen
    //oled.drawXBM(ScreenWidth_lastpixel - RegalMini_XMB_Width, (oled.getMaxCharHeight()+2-RegalMini_XMB_Height)/2, RegalMini_XMB_Width,RegalMini_XMB_Height,RegalMini_XMB);
    oled.drawXBMP(ScreenWidth_lastpixel - RegalMini_XMB_Width+1, 0, RegalMini_XMB_Width,RegalMini_XMB_Height,RegalMini_XMB);
    
    //Print one Value (complete display)
    oled.setFontPosBottom(); // text is build up from bottom of the font (bottom line)
    oled.setFont(u8g2_font_helvB24_tn);
    valueDescent = oled.getDescent(); 
    oled.setCursor(1,ScreenHeight_lastpixel);
    oled.print(value, 1);    

    //Print unit of the value
    oled.setFont(u8g2_font_helvB10_tr);
    unitDescent = oled.getDescent(); 
    oled.setFontPosBottom(); // text is build up from bottom of the font (bottom line)
    oled.setCursor(ScreenWidth_lastpixel - oled.getStrWidth(unit.c_str()),ScreenHeight_lastpixel + (valueDescent-unitDescent)); // LEvel unit and value (due to different Descent) 
    oled.print(unit);   
  }
  while (oled.nextPage());

**Comment by olikraus** [comment]
First it is important to use the font which includes just the glyphs required for your project, but "tr" is probably a good choice for your project.

U8g2 already removes unused functions. For example, the circle draw functions are not included, because you don't call them.
If you already occupy 10K, then also most the code comes from arduino. For example "print" is not a u8g2 function but comes from Arduino environment and is just reused in u8g2.

There are still some optimizations possible, please have a look into u8g2.h for this. There are pros and cons for this, but for example if you don't require font rotation, then you could save some bytes by commenting the corresponding define:

https://github.com/olikraus/u8g2/blob/b42953b87a60c10d0e0887f9259537a93f122aaf/csrc/u8g2.h#L124-L131

**Comment by RobLeop** [comment]
Hello olikraus,
thanks a lot for that precise answer!

I have another question which could fit in this topic: I like to add 3 letters into the type u8g2_font_helvB24_tn.
Could you describe how you build the fonts used and how can I modify them?

With this knowledge I could also reduce the size of some used fonts.
 
Have a great day!

**Reply by olikraus** [reply]
You can generate the font by yourself and add the required chars. First: Take the source file from here: https://github.com/olikraus/u8g2/blob/master/tools/font/bdf/helvB24.bdf

Second: Follow the instructions from the FAQ:
[ execute bdfconv ](https://github.com/olikraus/u8g2/blob/9ab4a1ecb8806da95fa24aa658b28dd7733974e4/doc/faq.txt#L252-L276)

There is especially a helper tool which helps to find the correct commindline options for bdfconv: https://stncrn.github.io/u8g2-unifont-helper/
